Feature metadata extension (EXT_feature_metadata) (#3)
CesiumGS/glTF: extensions/2.0/Vendor/EXT_feature_metadata/1.0.0/README.md

<!-- omit in toc -->
## Optional vs. Required

This extension is optional, meaning it should be placed in the `extensionsUsed` list, but not in the `extensionsRequired` list.

<!-- omit in toc -->
## Contents

- [Overview](#overview)
- [Feature Identification](#feature-identification)
  - [Feature ID Vertex Attributes](#feature-id-vertex-attributes)
    - [Feature ID Accessors](#feature-id-accessors)
    - [Implicit Feature IDs](#implicit-feature-ids)
  - [Feature ID Textures](#feature-id-textures)
  - [Feature ID Instance Attributes](#feature-id-instance-attributes)
- [Feature Metadata](#feature-metadata)
  - [Schemas](#schemas)
  - [Feature Tables](#feature-tables)
  - [Feature Textures](#feature-textures)
  - [Statistics](#statistics)
- [Examples](#examples)

## Overview

This extension allows offline batching of heterogeneous 3D features, such as different buildings in a city, for efficient streaming to a client for rendering and interaction. Efficiency comes from transferring multiple features in a single request and rendering them in the least number of draw calls necessary.

> Feature IDs enable individual features to be identified and updated at runtime, e.g., show/hide, highlight color, etc. Feature IDs may be assigned on a per-vertex, per-texel, or per-instance basis.

Reviewer: This first sentence feels a bit awkward, maybe reword. Also I'd say use `e.g.` OR use `etc.` but both together feels a bit much.
```suggestion
Feature IDs enable individual features to be identified and updated at runtime, e.g., show/hide, highlight color, etc. Feature IDs may be assigned on a per-vertex, per-texel, or per-instance basis.
```

> Feature IDs may be used to access metadata, such as passing a building's ID to get its address. Per-feature metadata is stored in a compact binary tabular format described in the [Cesium 3D Metadata Specification](https://github.com/CesiumGS/3d-tiles/blob/3d-tiles-next/specification/Metadata/README.md).

Reviewer: ```suggestion
Feature IDs may be used to access metadata, such as passing a building's ID to get its address. Feature metadata is stored in a compact binary tabular format described in the [Cesium 3D Metadata Specification](https://github.com/CesiumGS/3d-tiles/blob/3d-tiles-next/specification/Metadata/README.md).
```

![Building Example](figures/feature-table-buildings.png)

> > In the diagram above, a glTF consists of two houses batched together into a single primitive. A feature ID attribute on the primitive indicates that all of the vertices making up the first house have a feature ID of 0, while all vertices making up the second house have the feature ID 1. The feature ID is then used to access the building's metadata from the feature table.

Reviewer: just wondering, why is this a quote block? do we have other examples of using this for explanations?

> This extension also defines an alternate form of metadata storage that uses textures to store values directly. This format is especially useful when texture mapping high frequency data, like material properties, to less detailed 3D surfaces. Metadata textures enable new styling and analytics capabilities, and complement glTF PBR textures.

Reviewer: ```suggestion
This extension also defines an alternate form of metadata storage that uses textures to store values directly. This format is especially useful when texture mapping high frequency data, such as material properties, to less detailed 3D surfaces. Metadata textures enable new styling and analytics capabilities, and complement glTF PBR textures.
```

> This extension also defines an alternate form of metadata storage that uses textures to store values directly. This format is especially useful when texture mapping high frequency data, like material properties, to less detailed 3D surfaces. Metadata textures enable new styling and analytics capabilities, and complement glTF PBR textures.

Reviewer: the phrase "alternate form of metadata storage" sounds ambiguous to me, hard to tell "alternative to _what_?" Also we do use the term "raster format" in the core metadata spec now so calling it "form of metadata storage" feels a bit inconsistent.

> See [Use Cases](#use-cases) for a full list of use cases for this extension.

Reviewer: This is a dead hyperlink. Either add the section or remove this line.

### Feature ID Instance Attributes

Feature IDs may also be assigned to individual instances when using the [`EXT_mesh_gpu_instancing` extension](../../EXT_mesh_gpu_instancing/README.md). This works the same way as assigning feature IDs to vertices. Feature IDs may be stored in accessors or generated implicitly.

```jsonc
{
  "nodes": [
    {
      "mesh": 0,
      "extensions": {
        "EXT_mesh_gpu_instancing": {
          "attributes": {
            "TRANSLATION": 0,
            "ROTATION": 1,
            "SCALE": 2,
            "_FEATURE_ID_0": 
          },
          "extensions": {
            "EXT_feature_metadata": {
              "featureIdAttributes": [
                {
                  "featureTable": "trees",
                  "featureIds": {
                    "attribute": "_FEATURE_ID_0"
                  }
                }
              ]
            }
          }
        }
      }
    }
  ]
}
```

> ## Feature Metadata

Reviewer: Just wondering, why do we discuss feature metadata before feature identification? While I guess it's somewhat arbitrary, to me it makes more sense if you explain what metadata is, _then_ describe how to reference it via identification.

Author: It could be argued that feature identification is more core to the extension because it allows individual features to be identified, which is a perquisite to assigning metadata. There are other operations you can do on features besides using metadata (show/hide/highlight color).

Author: But I agree, it's somewhat arbitrary. This organization just felt right this go around, especially in lieu of the updated positioning in the overview.

Each buffer view `byteOffset` must be aligned to a multiple of 8 bytes. For a GLB file, this is measured relative to the beginning of the file. For a glTF + BIN file, this is relative to the beginning of the BIN file.

### Feature Textures

Feature textures (not to be confused with [Feature ID Textures](#feature-id-texture)) are an alternate form of metadata storage that uses textures rather than parallel arrays to store values. Feature textures are accessed directly by texture coordinates, rather than feature IDs. Feature textures are especially useful when texture mapping high frequency data to less detailed 3D surfaces.

Feature textures use the [Raster Format](https://github.com/CesiumGS/3d-tiles/tree/3d-tiles-next/specification/Metadata/1.0.0#raster-format) of the [Cesium 3D Metadata Specification](https://github.com/CesiumGS/3d-tiles/blob/3d-tiles-next/specification/Metadata/README.md) with a few additional constraints:

* A scalar property cannot be encoded into multiple channels. For example, it is not possible to encode a `UINT32` property in an `RGBA8` texture.
* Components of fixed-length array properties must be separate channels within the same texture.

> * Variable-length arrays are not supported

Reviewer: For consistency, everything in the list should be a sentence
```suggestion
* Variable-length arrays are not supported.
```

Additionally, the data type and bit depth of the image must be compatible with the property type. An 8-bit per pixel RGB image is only compatible with `UINT8` or normalized `UINT8` properties, and array properties thereof with three components or less. Likewise, a floating point property requires a floating point-compatible image format like KTX2 which may require additional extensions.

Feature textures are defined with the following steps:

> * A class is defined in the root `EXT_feature_metadata` extension object. This is used to describe the metadata in the texture.
> * A feature texture is defined in the root `EXT_feature_metadata.featureTextures` object. This must reference the class ID defined in step 1.
> * A feature texture is associated with a primitive by listing the feature texture ID in the `primitive.EXT_feature_metadata.featureTextures` array.

Reviewer: ```suggestion
1. A class is defined in the root `EXT_feature_metadata` extension object. This is used to describe the metadata in the texture.
2. A feature texture is defined in the root `EXT_feature_metadata.featureTextures` object. This must reference the class ID defined in step 1.
3. A feature texture is associated with a primitive by listing the feature texture ID in the `primitive.EXT_feature_metadata.featureTextures` array.
```
